consensus/src/lib.rs

```diff
@@ -1,27 +1,1101 @@
+            self.check_block_fee(transactions)?;
+            self.check_tx_inputs(transactions)?;
+        }
+        self.check_tx_outputs(transactions)?;
+        self.connect_transactions(&block)?;
+
+        match &new_tip_block_index.get_prev_block_id() {
+            Some(prev_block_id) => {
+                // To connect a new block we should set-up the next_block_id field of the previous block index
+                let mut prev_block = self
+                    .db_tx
+                    .get_block_index(prev_block_id)?
+                    .expect("Can't get block index. Inconsistent DB");
+                prev_block.set_next_block_id(block.get_id());
+                self.db_tx
+                    .set_block_index(&prev_block)
+                    .expect("Can't set block index. Inconsistent DB");
+            }
+            None => {
+                if best_block_id.is_some() {
+                    panic!("Failed to read block")
+                }
+            }
+        }
+        self.db_tx.set_block_index(new_tip_block_index)?;
+        self.db_tx.set_best_block_id(new_tip_block_index.get_block_id())?;
+        Ok(())
+    }
+
+    fn disconnect_tip(&mut self, block_index: &mut BlockIndex) -> Result<(), BlockError> {
+        assert_eq!(
+            &self.db_tx.get_best_block_id().ok().flatten().expect("Only fails at genesis"),
+            block_index.get_block_id()
+        );
+        let block = self.get_block_from_index(block_index)?.expect("Inconsistent DB");
+        // Disconnect transactions
+        self.disconnect_transactions(block.transactions())?;
+        self.db_tx.set_best_block_id(
+            block_index.get_prev_block_id().as_ref().ok_or(BlockError::Unknown)?,
+        )?;
+        // Disconnect block
+        let mut ancestor = self.get_previous_block_index(block_index.get_block_id())?;
+        ancestor.unset_next_block_id();
+        self.db_tx.set_block_index(&ancestor)?;
+        Ok(())
+    }
+
+    fn try_connect_genesis_block(
+        &mut self,
+        new_block_index: &BlockIndex,
+        best_block_id: &Option<Id<Block>>,
+    ) -> Result<Option<BlockIndex>, BlockError> {
+        if best_block_id.is_none() && new_block_index.is_genesis(self.chain_config) {
+            self.connect_tip(new_block_index, best_block_id)?;
+            self.db_tx
+                .set_best_block_id(new_block_index.get_block_id())
+                .map_err(BlockError::from)?;
+            return Ok(Some(new_block_index.clone()));
+        }
+        Ok(None)
+    }
+
+    fn activate_best_chain(
+        &mut self,
+        new_block_index: BlockIndex,
+        best_block_id: Option<Id<Block>>,
+    ) -> Result<Option<BlockIndex>, BlockError> {
+        let connected_genesis = self.try_connect_genesis_block(&new_block_index, &best_block_id)?;
+        if connected_genesis.is_some() {
+            return Ok(connected_genesis);
+        }
+
+        if let Some(best_block_id) = best_block_id {
```

Comment:
Like mentioned in the call. At this point, best_block_id MUST be set. So you should either expect on it or extract it from the previous step. This if is just a recipe for bugs.

Reply:
Fixed here and also fixed in `add_to_block_index`